### PR TITLE
FIX: mnist example failing in image preprocessing

### PR DESCRIPTION
**Pull request by jimthompson5802**
# Code Pull Requests

When running the MNIST example, this error occurs:
```
Traceback (most recent call last):
  File "/opt/project/examples/mnist/simple_model_training.py", line 35, in <module>
    ) = model.train(
  File "/opt/project/ludwig/api.py", line 488, in train
    self.model = LudwigModel.create_model(self.config,
  File "/opt/project/ludwig/api.py", line 1596, in create_model
    return ECD(
  File "/opt/project/ludwig/models/ecd.py", line 49, in __init__
    self.input_features.update(build_inputs(input_features_def))
  File "/opt/project/ludwig/models/ecd.py", line 307, in build_inputs
    input_features[input_feature_def[NAME]] = build_single_input(
  File "/opt/project/ludwig/models/ecd.py", line 336, in build_single_input
    input_feature_obj = input_feature_class(input_feature_def, encoder_obj)
  File "/opt/project/ludwig/features/image_feature.py", line 446, in __init__
    self.encoder_obj = self.initialize_encoder(feature)
  File "/opt/project/ludwig/features/base_feature.py", line 111, in initialize_encoder
    return get_from_registry(self.encoder, self.encoder_registry)(
  File "/opt/project/ludwig/encoders/image_encoders.py", line 126, in __init__
    self.fc_stack = FCStack(
  File "/opt/project/ludwig/modules/fully_connected_modules.py", line 163, in __init__
    FCLayer(
  File "/opt/project/ludwig/modules/fully_connected_modules.py", line 57, in __init__
    fc = Linear(
  File "/usr/local/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 85, in __init__
    self.weight = Parameter(torch.empty((out_features, in_features), **factory_kwargs))
RuntimeError: Trying to create tensor with negative dimension -320: [128, -320]
```

This PR changes how images are read in when inferring the height/width of images.

### CODE REVIEW

**Comment by jimthompson5802** [comment]
Marking PR ready for review.   Summary of changes:
* Fix error when inferring image height/width
* Add unit test to for inferring image metadata
* Fix memory leak in batch_evaluation process

File: ludwig/features/base_feature.py [hunk 1]
```diff
@@ -209,13 +209,14 @@ def train_loss(self, targets: Tensor, predictions: Dict[str, Tensor]):
     def eval_loss(self, targets: Tensor, predictions: Dict[str, Tensor]):
         loss_class = type(self.train_loss_function)
         prediction_key = LOSS_INPUTS_REGISTRY[loss_class]
-        return self.eval_loss_function(predictions[prediction_key], targets)
+        return self.eval_loss_function(predictions[prediction_key].detach(),
```
Comment (justinxzhao) [inline ludwig/features/base_feature.py:212]: I'm wondering if `.clone().detach()` would make a difference regarding the issue where the reported combined loss isn't the same as the reported loss for the output feature.

**Review by justinxzhao — APPROVED** [review]

**Review by ShreyaR — APPROVED** [review]
Thanks a lot for the fix @jimthompson5802!

File: ludwig/features/base_feature.py (hunk 1 above)
Comment (jimthompson5802) [reply ludwig/features/base_feature.py:212]: @justinxzhao   thank you for the suggestion.  Tried `clone().detach()`.  No change in output
```
╒═════════╤════════╤════════════╤═════════════╕
│ label   │   loss │   accuracy │   hits_at_k │
╞═════════╪════════╪════════════╪═════════════╡
│ train   │ 0.1200 │     0.9786 │      0.9981 │
├─────────┼────────┼────────────┼─────────────┤
│ test    │ 0.1173 │     0.9793 │      0.9985 │
╘═════════╧════════╧════════════╧═════════════╛
╒════════════╤════════╕
│ combined   │   loss │
╞════════════╪════════╡
│ train      │ 0.0713 │
├────────────┼────────┤
│ test       │ 0.0636 │
╘════════════╧════════╛
```
As I recall the fix in TF version, it required rationalizing how the loss was captured between two different ways.  I'll have to go back the the original fix to recall what was done.

For the moment, I'm leaving the code as-is.

Comment (justinxzhao) [reply ludwig/features/base_feature.py:212]: Thanks for checking on this! Perhaps it's a good thing that we're seeing the same behavior as the TF version.

**Pull request merged**